rest-server/src/controllers/job.js

```diff
@@ -111,35 +89,14 @@ const get = (req, res) => {
 /**
  * Submit or update job.
  */
-const update = (req, res) => {
+const update = (req, res, next) => {
   let name = req.job.name;
   let data = req.body;
   data.originalData = req.originalBody;
   data.userName = req.user.username;
   Job.prototype.putJob(name, data, (err) => {
     if (err) {
-      logger.warn('update job %s error\n%s', name, err.stack);
-      if (err.message === 'VirtualClusterNotFound') {
-        return res.status(500).json({
-          error: 'JobUpdateWithInvalidVirtualCluster',
-          message: `job update error: could not find virtual cluster ${data.virtualCluster}`,
-        });
-      } else if (err.message === 'NoRightAccessVirtualCluster') {
-        return res.status(401).json({
-          error: 'JobUpdateWithNoRightVirtualCluster',
-          message: `job update error: no virtual cluster right to access ${data.virtualCluster}`,
-        });
-      } else if (err.message === 'VirtualClusterNotFoundInDatabase') {
-        return res.status(404).json({
-          error: 'VirtualClusterNotFoundInDatabase',
-          message: `job update error: search virtual cluster from db failed`,
-        });
-      } else {
-        return res.status(500).json({
-          error: 'JobUpdateError',
-          message: err.message,
-        });
-      }
+      return next(createError.unknown(err));
```

Comment:
I think your method covers too many details.

Reply:
I suggest we split errors into two categories:

- Internal errors, which are raised by models and handled by controllers;
- Exposed errors, which will be returned to API callers.

Errors in the first category contain many details, for example HDFS error, launcher error, VC error, etc., which are not necessary to be exposed to API callers. After the processed by controllers, these internal errors will eventually be reorganized and translated into exposed errors, which contain meaningful information relevant to external developers.

Reply:
I think those errors have been handled in models instead of here.

Reply:
- `VirtualClusterNotFound` raised in https://github.com/Microsoft/pai/blob/qixcheng/rest-server-refine-error/job/rest-server/src/models/user.js#L204
- `NoRightAccessVirtualCluster` raised in https://github.com/Microsoft/pai/blob/qixcheng/rest-server-refine-error/job/rest-server/src/models/user.js#L216
- `VirtualClusterNotFoundInDatabase` covers too many details in https://github.com/Microsoft/pai/blob/qixcheng/rest-server-refine-error/job/rest-server/src/models/user.js#L208 so I throwed a `UnknownError` with HDFS Error message.